pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,335 @@
+			pvc := obj.(*corev1.PersistentVolumeClaim)
+			if IsPVCDataSourceRefKind(pvc, sourceKind) {
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			if isPVCPrimeDataSourceRefKind(pvc, sourceKind) {
+				owner := metav1.GetControllerOf(pvc)
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: owner.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			return nil
+		}),
+	); err != nil {
+		return err
+	}
+
+	mapDataSourceRefToPVC := func(obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{dataSourceRefField: getPopulatorIndexKey(obj.GetNamespace(), sourceKind, obj.GetName())}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return
+	}
+
+	if err := c.Watch(&source.Kind{Type: sourceType},
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateCommonPopulatorIndexes creates indexes used by all populators
+func CreateCommonPopulatorIndexes(mgr manager.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.PersistentVolumeClaim{}, dataSourceRefField, func(obj client.Object) []string {
+		pvc := obj.(*corev1.PersistentVolumeClaim)
+		dataSourceRef := pvc.Spec.DataSourceRef
+		if dataSourceRef != nil && dataSourceRef.APIGroup != nil &&
+			*dataSourceRef.APIGroup == cc.AnnAPIGroup && dataSourceRef.Name != "" {
+			return []string{getPopulatorIndexKey(obj.GetNamespace(), pvc.Spec.DataSourceRef.Kind, pvc.Spec.DataSourceRef.Name)}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcilerBase) getPVCPrime(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	pvcPrime := &corev1.PersistentVolumeClaim{}
+	pvcPrimeKey := types.NamespacedName{Namespace: pvc.Namespace, Name: PVCPrimeName(pvc)}
+	if err := r.client.Get(context.TODO(), pvcPrimeKey, pvcPrime); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) getPV(volName string) (*corev1.PersistentVolume, error) {
+	pv := &corev1.PersistentVolume{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: volName}, pv); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (r *ReconcilerBase) handleStorageClass(pvc *corev1.PersistentVolumeClaim) (bool, bool, error) {
+	if pvc.Spec.StorageClassName == nil {
+		return false, false, nil
```

Comment:
in this case dont we want to return ready true? currently we will just do nothing..

Reply:
Good point, will change that